File: golem-worker-executor-base/src/services/blob_store.rs
// Copyright 2024-2025 Golem Cloud
//
// Licensed under the Golem Source License v1.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://license.golem.cloud/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::path::{Path, PathBuf};
use std::sync::Arc;

use anyhow::anyhow;
use async_trait::async_trait;
use bincode::{Decode, Encode};

use golem_common::model::AccountId;

use golem_service_base::storage::blob::{BlobStorage, BlobStorageNamespace, ExistsResult};
use golem_wasm_rpc_derive::IntoValue;

/// Interface for storing blobs in a persistent storage.
#[async_trait]
pub trait BlobStoreService: Send + Sync {
    async fn clear(&self, account_id: AccountId, container_name: String) -> anyhow::Result<()>;

    async fn container_exists(
        &self,
        account_id: AccountId,
        container_name: String,
    ) -> anyhow::Result<bool>;

    async fn copy_object(
        &self,
        account_id: AccountId,
        source_container_name: String,
        source_object_name: String,
        destination_container_name: String,
        destination_object_name: String,
    ) -> anyhow::Result<()>;

    async fn create_container(
        &self,
        account_id: AccountId,
        container_name: String,
    ) -> anyhow::Result<()>;

    async fn delete_container(
        &self,
        account_id: AccountId,
        container_name: String,
    ) -> anyhow::Result<()>;

    async fn delete_object(
        &self,
        account_id: AccountId,
        container_name: String,
        object_name: String,
    ) -> anyhow::Result<()>;

    async fn delete_objects(
        &self,
        account_id: AccountId,
        container_name: String,
        object_names: Vec<String>,
    ) -> anyhow::Result<()>;

    async fn get_container(
        &self,
        account_id: AccountId,
        container_name: String,
    ) -> anyhow::Result<Option<u64>>;

    async fn get_data(
        &self,
        account_id: AccountId,
        container_name: String,
        object_name: String,
        start: u64,
        end: u64,
    ) -> anyhow::Result<Vec<u8>>;

    async fn has_object(
        &self,
        account_id: AccountId,
        container_name: String,
        object_name: String,
    ) -> anyhow::Result<bool>;

    async fn list_objects(
        &self,
        account_id: AccountId,
        container_name: String,
    ) -> anyhow::Result<Vec<String>>;

    async fn move_object(
        &self,
        account_id: AccountId,
        source_container_name: String,
        source_object_name: String,
        destination_container_name: String,
        destination_object_name: String,
    ) -> anyhow::Result<()>;

    async fn object_info(
        &self,
        account_id: AccountId,
        container_name: String,
        object_name: String,
    ) -> anyhow::Result<ObjectMetadata>;

    async fn write_data(
        &self,
        account_id: AccountId,
        container_name: String,
        object_name: String,
        data: Vec<u8>,
    ) -> anyhow::Result<()>;
}

pub struct DefaultBlobStoreService {
    blob_storage: Arc<dyn BlobStorage + Send + Sync>,
}

impl DefaultBlobStoreService {
    pub fn new(blob_storage: Arc<dyn BlobStorage + Send + Sync>) -> Self {
        Self { blob_storage }
    }
}

#[async_trait]
impl BlobStoreService for DefaultBlobStoreService {
    async fn clear(&self, account_id: AccountId, container_name: String) -> anyhow::Result<()> {
        self.blob_storage
            .delete_dir(
                "blob_store",
                "clear",
                BlobStorageNamespace::CustomStorage(account_id),
                Path::new(&container_name),
            )
            .await
            .map_err(|err| anyhow!(err))?;
        Ok(())
    }

    async fn container_exists(
        &self,
        account_id: AccountId,
        container_name: String,
    ) -> anyhow::Result<bool> {
        self.blob_storage
            .exists(
                "blob_store",
                "container_exists",
                BlobStorageNamespace::CustomStorage(account_id),
                Path::new(&container_name),
            )
            .await
            .map_err(|err| anyhow!(err))
            .map(|result| match result {
                ExistsResult::Directory => true,
                ExistsResult::File => false,
                ExistsResult::DoesNotExist => false,
            })
    }

    async fn copy_object(
        &self,
        account_id: AccountId,
        source_container_name: String,
        source_object_name: String,
        destination_container_name: String,
        destination_object_name: String,
    ) -> anyhow::Result<()> {
        self.blob_storage
            .copy(
                "blob_store",
                "copy_object",
                BlobStorageNamespace::CustomStorage(account_id),
                &Path::new(&source_container_name).join(&source_object_name),
                &Path::new(&destination_container_name).join(&destination_object_name),
            )
            .await
            .map_err(|err| anyhow!(err))
    }

    async fn create_container(
        &self,
        account_id: AccountId,
        container_name: String,
    ) -> anyhow::Result<()> {
        self.blob_storage
            .create_dir(
                "blob_store",
                "create_container",
                BlobStorageNamespace::CustomStorage(account_id),
                Path::new(&container_name),
            )
            .await
            .map_err(|err| anyhow!(err))?;
        Ok(())
    }

    async fn delete_container(
        &self,
        account_id: AccountId,
        container_name: String,
    ) -> anyhow::Result<()> {
        self.blob_storage
            .delete_dir(
                "blob_store",
                "delete_container",
                BlobStorageNamespace::CustomStorage(account_id),
                Path::new(&container_name),
            )
            .await
            .map_err(|err| anyhow!(err))?;
        Ok(())
    }

    async fn delete_object(
        &self,
        account_id: AccountId,
        container_name: String,
        object_name: String,
    ) -> anyhow::Result<()> {
        self.blob_storage
            .delete_dir(
                "blob_store",
                "delete_object",
                BlobStorageNamespace::CustomStorage(account_id),
                &Path::new(&container_name).join(&object_name),
            )
            .await
            .map_err(|err| anyhow!(err))?;
        Ok(())
    }

    async fn delete_objects(
        &self,
        account_id: AccountId,
        container_name: String,
        object_names: Vec<String>,
    ) -> anyhow::Result<()> {
        let paths: Vec<PathBuf> = object_names
            .iter()
            .map(|object_name| Path::new(&container_name).join(object_name))
            .collect();
        self.blob_storage
            .delete_many(
                "blob_store",
                "delete_objects",
                BlobStorageNamespace::CustomStorage(account_id),
                &paths,
            )
            .await
            .map_err(|err| anyhow!(err))
    }

    async fn get_container(
        &self,
        account_id: AccountId,
        container_name: String,
    ) -> anyhow::Result<Option<u64>> {
        self.blob_storage
            .get_metadata(
                "blob_store",
                "get_container",
                BlobStorageNamespace::CustomStorage(account_id),
                Path::new(&container_name),
            )
            .await
            .map_err(|err| anyhow!(err))
            .map(|result| result.map(|metadata| metadata.last_modified_at.to_millis()))
    }

    async fn get_data(
        &self,
        account_id: AccountId,
        container_name: String,
        object_name: String,
        start: u64,
        end: u64,
    ) -> anyhow::Result<Vec<u8>> {
        let data = self
            .blob_storage
            .get_raw_slice(
                "blob_store",
                "get_data",
                BlobStorageNamespace::CustomStorage(account_id),
                &Path::new(&container_name).join(&object_name),
                start,
                end,
            )
            .await
            .map_err(|err| anyhow!(err))?;

        match data {
            Some(data) => Ok(data.to_vec()),
            None => anyhow::bail!("Object does not exist"),
        }
    }

    async fn has_object(
        &self,
        account_id: AccountId,
        container_name: String,
        object_name: String,
    ) -> anyhow::Result<bool> {
        self.blob_storage
            .exists(
                "blob_store",
                "has_object",
                BlobStorageNamespace::CustomStorage(account_id),
                &Path::new(&container_name).join(&object_name),
            )
            .await
            .map_err(|err| anyhow!(err))
            .map(|result| match result {
                ExistsResult::Directory => false,
                ExistsResult::File => true,
                ExistsResult::DoesNotExist => false,
            })
    }

    async fn list_objects(
        &self,
        account_id: AccountId,
        container_name: String,
    ) -> anyhow::Result<Vec<String>> {
        self.blob_storage
            .list_dir(
                "blob_store",
                "list_objects",
                BlobStorageNamespace::CustomStorage(account_id),
                Path::new(&container_name),
            )
            .await
            .map_err(|err| anyhow!(err))
            .map(|paths| {
                paths
                    .iter()
                    .map(|path| path.file_name().unwrap().to_string_lossy().to_string())
                    .collect()
            })
    }

    async fn move_object(
        &self,
        account_id: AccountId,
        source_container_name: String,
        source_object_name: String,
        destination_container_name: String,
        destination_object_name: String,
    ) -> anyhow::Result<()> {
        self.blob_storage
            .r#move(
                "blob_store",
                "move_object",
                BlobStorageNamespace::CustomStorage(account_id),
                &Path::new(&source_container_name).join(&source_object_name),
                &Path::new(&destination_container_name).join(&destination_object_name),
            )
            .await
            .map_err(|err| anyhow!(err))
    }

    async fn object_info(
        &self,
        account_id: AccountId,
        container_name: String,
        object_name: String,
    ) -> anyhow::Result<ObjectMetadata> {
        match self
            .blob_storage
            .get_metadata(
                "blob_store",
                "object_info",
                BlobStorageNamespace::CustomStorage(account_id),
                &Path::new(&container_name).join(&object_name),
            )
            .await
            .map_err(|err| anyhow!(err))?
        {
            Some(metadata) => Ok(ObjectMetadata {
                name: object_name,
                container: container_name,
                created_at: metadata.last_modified_at.to_millis(),
                size: metadata.size,
            }),
            None => anyhow::bail!("Object does not exist"),
        }
    }

    async fn write_data(
        &self,
        account_id: AccountId,
        container_name: String,
        object_name: String,
        data: Vec<u8>,
    ) -> anyhow::Result<()> {
        self.blob_storage
            .put_raw(
                "blob_store",
                "write_data",
                BlobStorageNamespace::CustomStorage(account_id),
                &Path::new(&container_name).join(&object_name),
                &data,
            )
            .await
            .map_err(|err| anyhow!(err))
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode, IntoValue)]
pub struct ObjectMetadata {
    pub name: String,
    pub container: String,
    pub created_at: u64,
    pub size: u64,
}

#[cfg(test)]
mod tests {
    use test_r::test;

    use std::path::Path;
    use std::sync::Arc;

    use tempfile::TempDir;

    use golem_common::model::AccountId;

    use crate::services::blob_store::{BlobStoreService, DefaultBlobStoreService};
    use golem_service_base::storage::blob::fs::FileSystemBlobStorage;
    use golem_service_base::storage::blob::memory::InMemoryBlobStorage;

    async fn test_container_exists(blob_store: &impl BlobStoreService) {
        let account1 = AccountId {
            value: "account1".to_string(),
        };
        assert!(!blob_store
            .container_exists(account1.clone(), "container1".to_string())
            .await
            .unwrap());
        blob_store
            .create_container(account1.clone(), "container1".to_string())
            .await
            .unwrap();
        assert!(blob_store
            .container_exists(account1.clone(), "container1".to_string())
            .await
            .unwrap());
    }

    async fn test_container_delete(blob_store: &impl BlobStoreService) {
        let account1 = AccountId {
            value: "account1".to_string(),
        };
        blob_store
            .create_container(account1.clone(), "container1".to_string())
            .await
            .unwrap();
        blob_store
            .delete_container(account1.clone(), "container1".to_string())
            .await
            .unwrap();
        assert!(!blob_store
            .container_exists(account1.clone(), "container1".to_string())
            .await
            .unwrap());
    }

    async fn test_container_has_write_read_has(blob_store: &impl BlobStoreService) {
        let account1 = AccountId {
            value: "account1".to_string(),
        };

        blob_store
            .create_container(account1.clone(), "container1".to_string())
            .await
            .unwrap();
        assert!(!blob_store
            .has_object(
                account1.clone(),
                "container1".to_string(),
                "obj1".to_string()
            )
            .await
            .unwrap());

        let original_data = vec![1, 2, 3, 4];
        blob_store
            .write_data(
                account1.clone(),
                "container1".to_string(),
                "obj1".to_string(),
                original_data.clone(),
            )
            .await
            .unwrap();

        let read_data = blob_store
            .get_data(
                account1.clone(),
                "container1".to_string(),
                "obj1".to_string(),
                0,
                4,
            )
            .await
            .unwrap();

        assert_eq!(original_data, read_data);
        assert!(blob_store
            .has_object(
                account1.clone(),
                "container1".to_string(),
                "obj1".to_string()
            )
            .await
            .unwrap());
    }

    async fn test_container_list_copy_move_list(blob_store: &impl BlobStoreService) {
        let account1 = AccountId {
            value: "account1".to_string(),
        };

        blob_store
            .create_container(account1.clone(), "container1".to_string())
            .await
            .unwrap();
        blob_store
            .create_container(account1.clone(), "container2".to_string())
            .await
            .unwrap();

        assert!(blob_store
            .list_objects(account1.clone(), "container1".to_string(),)
            .await
            .unwrap()
            .is_empty());

        let original_data = vec![1, 2, 3, 4];
        blob_store
            .write_data(
                account1.clone(),
                "container1".to_string(),
                "obj1".to_string(),
                original_data.clone(),
            )
            .await
            .unwrap();

        blob_store
            .copy_object(
                account1.clone(),
                "container1".to_string(),
                "obj1".to_string(),
                "container1".to_string(),
                "obj2".to_string(),
            )
            .await
            .unwrap();

        let mut result = blob_store
            .list_objects(account1.clone(), "container1".to_string())
            .await
            .unwrap();

        result.sort();

        assert_eq!(result, vec!["obj1", "obj2"]);

        blob_store
            .move_object(
                account1.clone(),
                "container1".to_string(),
                "obj1".to_string(),
                "container2".to_string(),
                "obj3".to_string(),
            )
            .await
            .unwrap();

        assert_eq!(
            blob_store
                .list_objects(account1.clone(), "container1".to_string(),)
                .await
                .unwrap(),
            vec!["obj2"]
        );

        assert_eq!(
            blob_store
                .list_objects(account1.clone(), "container2".to_string(),)
                .await
                .unwrap(),
            vec!["obj3"]
        );
    }

    fn in_memory_blob_store() -> impl BlobStoreService {
        let blob_storage = Arc::new(InMemoryBlobStorage::new());
        DefaultBlobStoreService::new(blob_storage)
    }

    async fn fs_blob_store(path: &Path) -> impl BlobStoreService {
        let blob_storage = Arc::new(FileSystemBlobStorage::new(path).await.unwrap());
        DefaultBlobStoreService::new(blob_storage)
    }

    #[test]
    async fn test_container_exists_in_memory() {
        let blob_store = in_memory_blob_store();
        test_container_exists(&blob_store).await;
    }

    #[test]
    async fn test_container_exists_local() {
        let tempdir = TempDir::new().unwrap();
        let blob_store = fs_blob_store(tempdir.path()).await;
        test_container_exists(&blob_store).await;
    }

    #[test]
    async fn test_container_delete_in_memory() {
        let blob_store = in_memory_blob_store();
        test_container_delete(&blob_store).await;
    }

    #[test]
    async fn test_container_delete_local() {
        let tempdir = TempDir::new().unwrap();
        let blob_store = fs_blob_store(tempdir.path()).await;
        test_container_delete(&blob_store).await;
    }

    #[test]
    async fn test_container_has_write_read_has_in_memory() {
        let blob_store = in_memory_blob_store();
        test_container_has_write_read_has(&blob_store).await;
    }

    #[test]
    async fn test_container_has_write_read_has_local() {
        let tempdir = TempDir::new().unwrap();
        let blob_store = fs_blob_store(tempdir.path()).await;
        test_container_has_write_read_has(&blob_store).await;
    }

    #[test]
    async fn test_container_list_copy_move_list_in_memory() {
        let blob_store = in_memory_blob_store();
        test_container_list_copy_move_list(&blob_store).await;
    }

    #[test]
    async fn test_container_list_copy_move_list_local() {
        let tempdir = TempDir::new().unwrap();
        let blob_store = fs_blob_store(tempdir.path()).await;
        test_container_list_copy_move_list(&blob_store).await;
    }
}

File: sdks/ts/packages/golem-ts-sdk/src/fluent/blobstore.ts
// Copyright 2024-2026 Golem Cloud
//
// Licensed under the Golem Source License v1.1 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://license.golem.cloud/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Plain-async (no Effect) fluent wrapper around `wasi:blobstore/blobstore` +
// `wasi:blobstore/container` + `wasi:blobstore/types`. Ported from
// effect-golem's `Blobstore.ts` + `host/BlobstoreClient.ts`, de-Effect-ified:
// operations return `Promise`s and throw a typed `BlobstoreError` instead of
// failing an Effect.

import * as Blob from 'wasi:blobstore/blobstore';
import * as ContainerNS from 'wasi:blobstore/container';
import * as Types from 'wasi:blobstore/types';
import { compileSchema } from './schema/adapter';
import type { StandardSchemaV1 } from './schema/standardSchema';

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

const messageOf = (e: unknown): string => {
  if (e !== null && typeof e === 'object') {
    const obj = e as { message?: string };
    if (typeof obj.message === 'string') return obj.message;
  }
  if (typeof e === 'string') return e;
  return String(e);
};

/**
 * Raised when any host call into `wasi:blobstore/*` traps. The blobstore WIT
 * error type is a plain `string`, so {@link trace} carries it verbatim.
 */
export class BlobstoreError extends Error {
  override readonly name = 'BlobstoreError';
  readonly trace: string;
  readonly operation: string;
  constructor(
    readonly cause: unknown,
    operation: string,
  ) {
    const t = messageOf(cause);
    super(`BlobstoreError(${operation}): ${t}`);
    this.trace = t;
    this.operation = operation;
  }
}

const wrap = <A>(operation: string, fn: () => A): A => {
  try {
    return fn();
  } catch (cause) {
    if (cause instanceof BlobstoreError) throw cause;
    throw new BlobstoreError(cause, operation);
  }
};

// ---------------------------------------------------------------------------
// Schema codec — UTF-8 JSON over a Standard Schema
// ---------------------------------------------------------------------------

const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder('utf-8', { fatal: true });

const validateSync = <T>(schema: StandardSchemaV1, value: unknown): T => {
  const result = schema['~standard'].validate(value);
  if (result instanceof Promise) {
    throw new Error('Schema validation must be synchronous for blobstore payloads');
  }
  if (result.issues !== undefined) {
    const msg = result.issues.map((i) => i.message).join('; ');
    throw new Error(`Schema validation failed: ${msg}`);
  }
  return result.value as T;
};

const encodeValue = <T>(schema: StandardSchemaV1, value: T): Uint8Array =>
  wrap('schema.encode', () => {
    const validated = validateSync<T>(schema, value);
    return textEncoder.encode(JSON.stringify(validated));
  });

const decodeValue = <T>(schema: StandardSchemaV1, bytes: Uint8Array): T =>
  wrap('schema.decode', () => {
    const json = textDecoder.decode(bytes);
    const parsed: unknown = JSON.parse(json);
    return validateSync<T>(schema, parsed);
  });

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

/** Object identifier — `(containerName, objectName)` pair. */
export interface ObjectId {
  readonly container: string;
  readonly object: string;
}

/**
 * Container metadata. The WIT `created-at` `u64` is populated by the Golem
 * host as Unix milliseconds (it is actually `last_modified_at`; object stores
 * have no separate creation time). Exposed as both a JS `Date` and the raw
 * bigint.
 */
export interface ContainerMetadata {
  readonly name: string;
  readonly createdAt: Date;
  readonly createdAtMillis: bigint;
}

/** Object metadata. Same `created-at` caveat as {@link ContainerMetadata}. */
export interface ObjectMetadata {
  readonly name: string;
  readonly container: string;
  readonly createdAt: Date;
  readonly createdAtMillis: bigint;
  readonly size: bigint;
}

/**
 * Schema-typed view of a {@link Container}. Object bodies are JSON-encoded via
 * the supplied Standard Schema.
 */
export interface SchemaContainer<T> {
  getData(name: string): Promise<T>;
  writeData(name: string, value: T): Promise<void>;
  has(name: string): Promise<boolean>;
  info(name: string): Promise<ObjectMetadata>;
  delete(name: string): Promise<void>;
}

/**
 * Handle on an open blob container. Acquired via {@link createContainer},
 * {@link getContainer} or {@link getOrCreateContainer}. GC-managed lifetime.
 */
export interface Container {
  readonly name: string;

  /** Container metadata. */
  info(): Promise<ContainerMetadata>;
  /** Remove all objects, leaving the container empty. */
  clear(): Promise<void>;

  /**
   * Read an object's bytes. With no range, the whole object is read (with a
   * recovery retry for the host's inclusive/exclusive end divergence). With an
   * explicit `[start, end]` range the bytes are passed to the host verbatim.
   *
   * **Backend caveat.** The Golem host diverges across backends: in-memory +
   * filesystem treat `end` as exclusive (Rust range), S3 treats it as
   * inclusive. The WIT spec says inclusive. Ranged reads are not portable.
   */
  getData(name: string, start?: bigint, end?: bigint): Promise<Uint8Array>;
  /** Create or replace `name` with `data` (chunked at 4096 bytes per write). */
  writeData(name: string, data: Uint8Array): Promise<void>;

  /** True if the named object exists in this container. */
  has(name: string): Promise<boolean>;
  /** Metadata for the named object. Fails if the object does not exist. */
  objectInfo(name: string): Promise<ObjectMetadata>;
  /** Delete the named object. Does NOT fail if it does not exist. */
  delete(name: string): Promise<void>;
  /** Delete multiple objects. */
  deleteMany(names: readonly string[]): Promise<void>;

  /**
   * All object names in this container. The Golem host eagerly snapshots the
   * full name list when `list-objects` is called and pins it to the oplog;
   * this drains the whole snapshot (page size 256). Order is not guaranteed.
   */
  listObjects(): Promise<string[]>;

  /** Build a schema-typed view of this container. */
  forSchema<S extends StandardSchemaV1>(
    schema: S,
  ): SchemaContainer<StandardSchemaV1.InferOutput<S>>;
}

// ---------------------------------------------------------------------------
// Host plumbing helpers
// ---------------------------------------------------------------------------

const CHUNK_SIZE = 4096;
const LIST_PAGE_SIZE = 256n;

const consumeIncoming = (iv: Types.IncomingValue): Uint8Array =>
  wrap('incomingValueConsumeSync', () => iv.incomingValueConsumeSync());

const buildOutgoingValue = (bytes: Uint8Array): Types.OutgoingValue =>
  wrap('outgoingValueWriteBody', () => {
    const ov = Types.OutgoingValue.newOutgoingValue();
    const stream = ov.outgoingValueWriteBody();
    const total = bytes.length;
    let offset = 0;
    while (offset < total) {
      const remaining = total - offset;
      const chunkLen = remaining > CHUNK_SIZE ? CHUNK_SIZE : remaining;
      stream.blockingWriteAndFlush(bytes.subarray(offset, offset + chunkLen));
      offset += chunkLen;
    }
    return ov;
  });

const toContainerMetadata = (
  m: ContainerNS.ContainerMetadata | Types.ContainerMetadata,
): ContainerMetadata => ({
  name: m.name,
  createdAt: new Date(Number(m.createdAt)),
  createdAtMillis: m.createdAt,
});

const toObjectMetadata = (m: Types.ObjectMetadata): ObjectMetadata => ({
  name: m.name,
  container: m.container,
  createdAt: new Date(Number(m.createdAt)),
  createdAtMillis: m.createdAt,
  size: m.size,
});

const readRange = (handle: ContainerNS.Container, name: string, start: bigint, end: bigint): Uint8Array =>
  wrap('container.getData', () => {
    const iv = handle.getData(name, start, end);
    return consumeIncoming(iv);
  });

// ---------------------------------------------------------------------------
// SchemaContainer factory
// ---------------------------------------------------------------------------

const makeSchemaContainer = <T>(container: Container, schema: StandardSchemaV1): SchemaContainer<T> => ({
  async getData(name) {
    const bytes = await container.getData(name);
    return decodeValue<T>(schema, bytes);
  },
  async writeData(name, value) {
    await container.writeData(name, encodeValue(schema, value));
  },
  has(name) {
    return container.has(name);
  },
  info(name) {
    return container.objectInfo(name);
  },
  delete(name) {
    return container.delete(name);
  },
});

// ---------------------------------------------------------------------------
// Container factory
// ---------------------------------------------------------------------------

const makeContainer = (name: string, handle: ContainerNS.Container): Container => {
  const objectInfo = async (objectName: string): Promise<ObjectMetadata> => {
    const m = wrap('container.objectInfo', () => handle.objectInfo(objectName));
    return toObjectMetadata(m);
  };

  const self: Container = {
    name,
    async info() {
      const m = wrap('container.info', () => handle.info());
      return toContainerMetadata(m);
    },
    async clear() {
      wrap('container.clear', () => handle.clear());
    },
    async getData(objectName, start, end) {
      // Explicit range — pass through verbatim (backend semantics diverge).
      if (start !== undefined && end !== undefined) {
        return readRange(handle, objectName, start, end);
      }
      // Whole-object read. The WIT spec says `end` is inclusive, but the Golem
      // in-memory and filesystem backends treat it as exclusive. Tolerate both:
      const meta = await objectInfo(objectName);
      const size = meta.size;
      if (size === 0n) return new Uint8Array(0);
      const firstAttempt = readRange(handle, objectName, 0n, size - 1n);
      if (BigInt(firstAttempt.length) === size) return firstAttempt;
      // Backend treats `end` as exclusive — replay with end = size.
      return readRange(handle, objectName, 0n, size);
    },
    async writeData(objectName, data) {
      const ov = buildOutgoingValue(data);
      wrap('container.writeData', () => handle.writeData(objectName, ov));
    },
    async has(objectName) {
      return wrap('container.hasObject', () => handle.hasObject(objectName));
    },
    objectInfo,
    async delete(objectName) {
      wrap('container.deleteObject', () => handle.deleteObject(objectName));
    },
    async deleteMany(names) {
      wrap('container.deleteObjects', () => handle.deleteObjects([...names]));
    },
    async listObjects() {
      const iter = wrap('container.listObjects', () => handle.listObjects());
      const out: string[] = [];
      let done = false;
      while (!done) {
        const [names, end] = wrap('streamObjectNames.read', () =>
          iter.readStreamObjectNames(LIST_PAGE_SIZE),
        );
        out.push(...names);
        done = end;
      }
      return out;
    },
    forSchema<S extends StandardSchemaV1>(schema: S) {
      compileSchema(schema);
      return makeSchemaContainer<StandardSchemaV1.InferOutput<S>>(self, schema);
    },
  };
  return self;
};

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/** Create a new empty container. Fails if a container of the same name exists. */
export async function createContainer(name: string): Promise<Container> {
  const handle = wrap('createContainer', () => Blob.createContainer(name));
  return makeContainer(name, handle);
}

/** Open an existing container by name. Fails if it does not exist. */
export async function getContainer(name: string): Promise<Container> {
  const handle = wrap('getContainer', () => Blob.getContainer(name));
  return makeContainer(name, handle);
}

/** True if a container with the given name exists. */
export async function containerExists(name: string): Promise<boolean> {
  return wrap('containerExists', () => Blob.containerExists(name));
}

/**
 * Open the named container, creating it first if it does not exist.
 * `wasi:blobstore` has no atomic get-or-create primitive, so this optimistically
 * calls `createContainer` and, on failure, replays `getContainer` if
 * `containerExists` now reports the container present (collapses the TOCTOU
 * window). Otherwise the original create failure propagates.
 */
export async function getOrCreateContainer(name: string): Promise<Container> {
  try {
    return await createContainer(name);
  } catch (createErr) {
    let exists = false;
    try {
      exists = await containerExists(name);
    } catch {
      exists = false;
    }
    if (exists) return getContainer(name);
    throw createErr;
  }
}

/** Delete a container and all of its objects. */
export async function deleteContainer(name: string): Promise<void> {
  wrap('deleteContainer', () => Blob.deleteContainer(name));
}

/** Copy an object to the same or a different container. Overwrites the destination. */
export async function copyObject(src: ObjectId, dest: ObjectId): Promise<void> {
  wrap('copyObject', () => Blob.copyObject(src, dest));
}

/** Move (rename) an object. Overwrites the destination. */
export async function moveObject(src: ObjectId, dest: ObjectId): Promise<void> {
  wrap('moveObject', () => Blob.moveObject(src, dest));
}
